### PR TITLE
all service keys now have MTA metadata

### PR DESCRIPTION
**Pull request by ikasarov**
service keys recreate implemented for general case for keys with generated names
refactored BuildUndeploy and BuildDeploy steps, flowable diagrams

### CODE REVIEW

File: multiapps-controller-core/src/main/java/org/cloudfoundry/multiapps/controller/core/helpers/SystemParameters.java [hunk 1]
```diff
@@ -36,7 +35,7 @@ public class SystemParameters {
     private final String authorizationEndpoint;
     private final String deployServiceUrl;
     private final boolean reserveTemporaryRoutes;
-    private final Supplier<String> timestampSupplier;
+    private final String timestamp;
```
Comment (radito3) [inline multiapps-controller-core/src/main/java/org/cloudfoundry/multiapps/controller/core/helpers/SystemParameters.java:38]: Why is this changed from a generator of a timestamp to a concrete timestamp that will be from the creation of the `SystemParameters` object?

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/CreateServiceKeyStep.java [hunk 2]
```diff
@@ -46,9 +46,8 @@ protected StepPhase executeAsyncStep(ProcessContext context) throws Exception {
     private Optional<String> createServiceKey(CloudServiceKey serviceKeyToCreate, CloudControllerClient controllerClient,
                                               ProcessContext context) {
         try {
-            return controllerClient.createServiceKey(serviceKeyToCreate.getServiceInstance()
-                                                                       .getName(),
-                                                     serviceKeyToCreate.getName(), serviceKeyToCreate.getCredentials());
```
Comment (radito3) [inline multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/CreateServiceKeyStep.java:51]: Why are the credentials removed from the create call?

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/DetermineServiceCreateUpdateServiceActionsStep.java [hunk 3]
```diff
@@ -32,6 +32,7 @@
 import org.springframework.beans.factory.config.BeanDefinition;
 import org.springframework.context.annotation.Scope;
 import org.springframework.http.HttpStatus;
+import static org.cloudfoundry.multiapps.controller.process.steps.StepsUtil.canDeleteServiceKeys;
```
Comment (radito3) [inline multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/DetermineServiceCreateUpdateServiceActionsStep.java:35]: Please don't :smile: 
Just use a normal import and do `return StepsUtil.canDeleteServiceKeys(context)`

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/DetermineServiceDeleteActionsToExecuteStep.java [hunk 4]
```diff
@@ -15,6 +15,7 @@
 import org.cloudfoundry.multiapps.controller.process.util.ServiceAction;
 import org.cloudfoundry.multiapps.controller.process.util.ServiceDeletionActions;
 import org.cloudfoundry.multiapps.controller.process.variables.Variables;
+import static org.cloudfoundry.multiapps.controller.process.steps.StepsUtil.canDeleteServiceKeys;
```
Comment (radito3) [inline multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/DetermineServiceDeleteActionsToExecuteStep.java:18]: Same here

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/DetermineServiceKeysToRecreateStep.java [hunk 5]
```diff
@@ -1,20 +1,19 @@
 package org.cloudfoundry.multiapps.controller.process.steps;
 
+import static org.cloudfoundry.multiapps.controller.process.steps.StepsUtil.canDeleteServiceKeys;
```
Comment (radito3) [inline multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/DetermineServiceKeysToRecreateStep.java:3]: Again here

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/DetermineServiceKeysToRecreateStep.java [hunk 6]
```diff
@@ -102,8 +101,13 @@ private boolean areServiceKeysEqual(CloudServiceKey key1, CloudServiceKey key2)
...
 
-    private boolean shouldDelete(CloudServiceKey existingKey, List<CloudServiceKey> keys) {
-        return getWithName(keys, existingKey.getName()) == null;
+    public static boolean isMetadataEqual(Metadata metadataA, Metadata metadataB) {
+        if (metadataA == null || metadataB == null) {
+            return metadataA == null && metadataB == null;
+        }
+
+        return Objects.equals(metadataA.getAnnotations(), metadataB.getAnnotations())
+            && Objects.equals(metadataA.getLabels(), metadataB.getLabels());
```
Comment (radito3) [inline multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/DetermineServiceKeysToRecreateStep.java:110]: Isn't the `Metadata::equals` implemented the same way? 
Why can't we directly use `Objects.equals(metadataA, metadataB)`?

File: multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/variables/Variables.java [hunk 7]
```diff
@@ -390,12 +393,12 @@ public interface Variables {
...
                                                                                                      })
                                                                                                      .build();
-    Variable<List<DeployedMtaServiceKey>> SERVICE_KEYS_TO_DELETE = ImmutableJsonBinaryVariable.<List<DeployedMtaServiceKey>> builder()
-                                                                                              .name("serviceKeysToDelete")
-                                                                                              .type(new TypeReference<>() {
-                                                                                              })
-                                                                                              .defaultValue(Collections.emptyList())
-                                                                                              .build();
+    Variable<List<DeployedMtaServiceKey>> SERVICE_KEYS_TO_DELETE = ImmutableJsonStringListVariable.<DeployedMtaServiceKey> builder()
```
Comment (radito3) [inline multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/variables/Variables.java:396]: Why is this changed?
This is backwards incompatible.

File: pom.xml [hunk 8]
```diff
@@ -38,7 +38,7 @@
         <jackson.databind.version>2.15.0</jackson.databind.version>
         <liquibase.version>4.22.0</liquibase.version>
         <liquibase-slf4j.version>4.1.0</liquibase-slf4j.version>
-        <cloudfoundry-client.version>2.32.0</cloudfoundry-client.version>
+        <cloudfoundry-client.version>2.33.0-SNAPSHOT</cloudfoundry-client.version>
```
Comment (radito3) [inline pom.xml:41]: Reminder to release the cf java client before merging this

File: multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/OperationsApiServiceImpl.java [hunk 9]
```diff
@@ -250,6 +251,8 @@ private Operation addServiceParameters(Operation operation, String spaceGuid, St
         parameters.put(Variables.ORGANIZATION_GUID.getName(), organization.getMetadata()
                                                                           .getGuid()
                                                                           .toString());
+        parameters.put(Variables.TIMESTAMP.getName(), DateTimeFormatter.ofPattern("yyyyMMddHHmmss")
```
Comment (radito3) [inline multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/OperationsApiServiceImpl.java:254]: Why this format?

File: multiapps-controller-core/src/main/java/org/cloudfoundry/multiapps/controller/core/cf/metadata/util/MtaMetadataUtil.java [hunk 10]
```diff
@@ -45,6 +47,7 @@ public static Metadata getMetadataWithoutMtaFields(Metadata metadata) {
                        .from(metadata)
                        .label(MtaMetadataLabels.MTA_ID, null)
                        .label(MtaMetadataLabels.MTA_NAMESPACE, null)
+                       .label(MtaMetadataLabels.SPACE_GUID, null)
```
Comment (radito3) [inline multiapps-controller-core/src/main/java/org/cloudfoundry/multiapps/controller/core/cf/metadata/util/MtaMetadataUtil.java:50]: Why is the space guid added and/or needed?

File: multiapps-controller-core/src/main/java/org/cloudfoundry/multiapps/controller/core/helpers/SystemParameters.java (hunk 1 above)
Comment (ikasarov) [reply multiapps-controller-core/src/main/java/org/cloudfoundry/multiapps/controller/core/helpers/SystemParameters.java:38]: this is changed, so that the timestamp will be immutable for the whole deployment (it's now generated at the start of the operation in [OperationsApiServiceImpl.addServiceParameters()) - before that the timestamp was generated twice for Blue-Green processes (in CollectBLueGreenSystemParametersStep and in the CollectSystemParametersStep after continue)

File: multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/OperationsApiServiceImpl.java (hunk 9 above)
Comment (ikasarov) [reply multiapps-controller-web/src/main/java/org/cloudfoundry/multiapps/controller/web/api/impl/OperationsApiServiceImpl.java:254]: It's the same format as before, taken from CollectSystemParametersStep

File: multiapps-controller-core/src/main/java/org/cloudfoundry/multiapps/controller/core/cf/metadata/util/MtaMetadataUtil.java (hunk 10 above)
Comment (ikasarov) [reply multiapps-controller-core/src/main/java/org/cloudfoundry/multiapps/controller/core/cf/metadata/util/MtaMetadataUtil.java:50]: It's needed for the DetectDeployedMtaStep as serviceKeys don't have a filter by spaceGuid in the v3 api - we add that as a label and select by that later
